File: Java/Graph/Detect cycle.java


    // undirected graph
    static List<Integer> graph[];
    static boolean vis[];
    static int parent[];
    static int n, m;
    static int star_cycle = -1, end_cycle = -1;


    static boolean dfs(int node, int par){
        vis[node] = true;
        parent[node] = par;

        for (int next : graph[node]) {
            if(next == par) continue;
            if(vis[next]){
                star_cycle = next;
                end_cycle = node;
                return true;
            }
            if(!vis[next] && dfs(next, node)){
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) throws IOException {
        n = nextInt();
        m = nextInt();

        graph = new List[n + 1];
        vis = new boolean[n + 1];
        parent = new int[n + 1];

        for (int i = 0; i < n + 1; i++) 
            graph[i] = new ArrayList<>();

        int a , b;
        for (int i = 0; i < m; i++) {
            a = nextInt();
            b = nextInt();
            graph[a].add(b);
            graph[b].add(a);
        }

        for (int i = 1; i < n + 1; i++) {
            if(!vis[i] && dfs(i, -1)){
                break;
            }
        }

        if(star_cycle == -1){
            sa.println("No hay ciclo");
        }else{
            sa.println("si hay ciclo");
            ArrayList<Integer> path = new ArrayList<>();
            path.add(end_cycle);
            for (int i = end_cycle; i != star_cycle; i=parent[i]) {
                path.add(parent[i]);
            }
            path.add(end_cycle);
            
            sa.println(path.size());
            for (int k : path) {
                sa.print(k+" ");
            }

        }

        sa.close();
    }


    //########################################################################################

    // directed graph

    
    static List<Integer> graph[];
    static int parent[];
    static int color[];
    static int n, m;
    static int cycle_start = -1, cycle_end = -1;

    static boolean dfs(int node) {
        color[node] = 1;
        for (int next : graph[node]) {
            if (color[next] == 0) {
                parent[next] = node;
                if (dfs(next)) {
                    return true;
                }
            } else if (color[next] == 1) {
                cycle_end = node;
                cycle_start = next;
                return true;
            }
        }
        color[node] = 2;
        return false;
    }

    public static void main(String[] args) throws IOException {
        n = nextInt();
        m = nextInt();

        graph = new List[n + 1];
        parent = new int[n + 1];
        color = new int[n + 1];

        for (int i = 0; i < n + 1; i++) {
            graph[i] = new ArrayList<>();
        }

        int a, b;
        for (int i = 0; i < m; i++) {
            a = nextInt();
            b = nextInt();
            graph[a].add(b);
        }

        for (int i = 1; i < n + 1; i++) {
            if (color[i] == 0 && dfs(i)) {
                break;
            }
        }
        
        if (cycle_start == -1) {
            sa.println("No hay ciclo");
        } else {
            sa.println("Si hay ciclo");
        }

        sa.close();
    }

File: cpp/Graph/Detect cycle.cpp

//directed graph
const int MAXN = 2e5 + 5;
vi g[MAXN];
int color[MAXN];
int in_cycle[MAXN];
int par[MAXN];

void dfs(int u){
    if(color[u] == 2) return;
    color[u] = 1;
    for(int v: g[u]){
        if(color[v] == 0){
            par[v] = u;
            dfs(v);
        }else if(color[v] == 1){
            //traverse the cycle
            int ini = u, fin = v;
            int cur = u;
            in_cycle[cur] = 1;
            while(cur != v){
                cur = par[cur];
                in_cycle[cur] = 1;
            }
            in_cycle[v] = 1;
        }
    }
    color[u] = 2;
}

/*
    forab(i, 1, n + 1) {
        in_cycle[i] = 0;
        color[i] = 0;
    }
    forab(i, 1, n + 1) {
        if(color[i] == 0) dfs(i);
    }
*/

//undirected graph
const int MAXN = 2e5 + 5;
int color[MAXN];
int mark[MAXN];
int par[MAXN];
vi g[MAXN];

void dfs(int u, int p, int &cy){
    if(color[u] == 2){
        return;
    }
    if(color[u] == 1){
        cy++;
        int cur = p;
        mark[cur] = cy;
        while(cur != u){
            cur = par[cur];
            mark[cur] = cy;
        }
        return;
    }
    par[u] = p;
    color[u] = 1;
    for(int v: g[u]){
        if(v == par[u]) continue;
        dfs(v, u, cy);
    }
    color[u] = 2;
}

/*
    forab(i, 1, n + 1){
        color[i] = 0;
        par[i] = i;
        mark[i] = -1;
    }
    int cycles = 0;
    dfs(1, 1, cycles);

    forab(i, 1, n + 1){
        if(mark[i] != -1){
            found = true;
        }
    }
*/
